src/Plugin/GraphQL/Fields/FieldPluginBase.php

```diff
@@ -64,29 +65,42 @@ public function getDefinition() {
    * {@inheritdoc}
    */
   public function resolve($value, array $args, ResolveContext $context, ResolveInfo $info) {
+    $definition = $this->getPluginDefinition();
+
     // If not resolving in a trusted environment, check if the field is secure.
     if (!$context->getGlobal('development', FALSE) && !$context->getGlobal('bypass field security', FALSE)) {
-      $definition = $this->getPluginDefinition();
       if (empty($definition['secure'])) {
         throw new \Exception(sprintf("Unable to resolve insecure field '%s'.", $info->fieldName));
       }
     }
 
-    return $this->resolveDeferred([$this, 'resolveValues'], $value, $args, $context, $info);
+    foreach ($definition['contextual_arguments'] as $argument) {
+      if (array_key_exists($argument, $args) && !is_null($args[$argument])) {
+        $context->setContext($argument, $args[$argument], $info);
+      }
+      else {
+        $args[$argument] = $context->getContext($argument, $info);
+      }
+    }
+
+    return \Drupal::service('graphql.language_context')
```

Comment:
Let's use setter injection for this (I understand why you don't want to use constructor injection here). Alternatively, we could put it into the ResolveContext.